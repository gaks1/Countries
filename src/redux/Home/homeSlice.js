import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const url = 'https://restcountries.com/v3.1/all';

const initialState = {
  countries: [],
  isLoading: false,
  fetched: false,
  error: null,
};

export const fetchCountries = createAsyncThunk('home/fetchCountries', async (_, thunkAPI) => {
  try {
    const response = await fetch(`${url}`);
    const data = await response.json();
    return data;
  } catch (error) {
    return thunkAPI.rejectWithValue(error.response.data);
  }
});

const homeSlice = createSlice({
  name: 'home',
  initialState,
  reducers: {},
  extraReducers: {
    [fetchCountries.pending]: (state) => ({ ...state, isLoading: true }),
    [fetchCountries.fulfilled]: (state, action) => ({
      ...state,
      isLoading: false,
      fetched: true,
      countries: action.payload,
    }),
    [fetchCountries.rejected]: (state, action) => ({
      ...state,
      isLoading: false,
      error: action.payload,
    }),
  },

});

export default homeSlice.reducer;
