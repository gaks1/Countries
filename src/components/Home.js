import { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { fetchCountries } from '../redux/Home/homeSlice';

const Home = () => {
  const {
    fetched, isLoading, countries,
  } = useSelector((state) => state.home);
  const dispatch = useDispatch();

  useEffect(() => {
    if (!fetched) {
      dispatch(fetchCountries());
    }
  }, [dispatch, fetched]);

  if (isLoading) {
    return (
      <p>Loading...</p>
    );
  }

  return (
    <div className="home">
      <div>
        <p>LIST OF COUNTRIES</p>
      </div>
      <ul>
        {countries.map((country) => (
          <Link key={country.name.common} to={`/Details/${country.name.common}`}>
            <li key={country.name.common}>
              <span className="material-symbols-outlined">
                arrow_circle_right
              </span>
              <div>
                <img src={country.flags.png} alt={country.flags.alt} />
              </div>
              <div>
                <p>
                  {country.name.common}
                </p>
                <p>
                  LAT:
                  {' '}
                  {country.latlng[0]}
                </p>
                <p>
                  LONG:
                  {' '}
                  {country.latlng[1]}
                </p>
              </div>
            </li>

          </Link>
        ))}
      </ul>
    </div>
  );
};

export default Home;
