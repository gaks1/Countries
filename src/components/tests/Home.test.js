import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import {
  BrowserRouter as Router,
} from 'react-router-dom';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Home from '../Home';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Home', () => {
  const initialState = {
    home: {
      countries: [
        {
          name: {
            common: 'Afghanistan',
          },
          flags: {
            png: 'https://restcountries.com/data/afg.svg',
            alt: 'flag',
          },
          latlng: [20, 20],
        },
        {
          name: {
            common: 'Albania',
          },
          flags: {
            png: 'https://restcountries.com/data/alb.svg',
            alt: 'flag',
          },
          latlng: [20, 20],
        },
      ],
      fetched: false,
      isLoading: false,
    },
  };

  const store = mockStore(initialState);

  test('renders without crashing', () => {
    const { getByText } = render(
      <Provider store={store}>
        <Router>
          <Home />
        </Router>
      </Provider>,
    );
    expect(getByText('Albania')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Home />
          </Router>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
