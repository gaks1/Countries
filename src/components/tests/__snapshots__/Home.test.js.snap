// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
<div
  className="home"
>
  <nav
    className="navbar"
  >
    <div>
      <a
        aria-current="page"
        className="links active"
        href="/"
        onClick={[Function]}
      >
        Countries
      </a>
    </div>
    <div
      className="header"
    >
      <div
        className="headerbutton"
      >
        <button
          onClick={[Function]}
          type="button"
        >
          <span
            className="material-symbols-outlined"
          >
            search
          </span>
        </button>
      </div>
    </div>
  </nav>
  <div
    className="divheader"
  >
    <div
      className="divtitle"
    >
      <div
        className="imageContainer"
      >
        <img
          alt=""
          src="Global-Map-Transparent-700x416.png"
        />
      </div>
      <div
        className="textContainer"
      >
        <h1>
          COUNTRIES AROUND
          <br />
          THE WORLD
        </h1>
      </div>
    </div>
  </div>
  <div
    className="separator"
  >
    <p
      className="septext"
    >
      LIST OF COUNTRIES
    </p>
  </div>
  <ul
    className="ulCountries"
  >
    <a
      className="link"
      href="/Details/Afghanistan"
      onClick={[Function]}
    >
      <li
        className="liCountry"
      >
        <span
          className="material-symbols-outlined arrowCircleRight"
        >
          arrow_circle_right
        </span>
        <div
          className="countryImage"
        >
          <img
            alt="flag"
            src="https://restcountries.com/data/afg.svg"
          />
        </div>
        <div
          className="countryDetails"
        >
          <p
            className="liname"
          >
            Afghanistan
          </p>
          <p
            className="licood"
          >
            LAT:
             
            20
          </p>
          <p
            className="licood"
          >
            LONG:
             
            20
          </p>
        </div>
      </li>
    </a>
    <a
      className="link"
      href="/Details/Albania"
      onClick={[Function]}
    >
      <li
        className="liCountry"
      >
        <span
          className="material-symbols-outlined arrowCircleRight"
        >
          arrow_circle_right
        </span>
        <div
          className="countryImage"
        >
          <img
            alt="flag"
            src="https://restcountries.com/data/alb.svg"
          />
        </div>
        <div
          className="countryDetails"
        >
          <p
            className="liname"
          >
            Albania
          </p>
          <p
            className="licood"
          >
            LAT:
             
            20
          </p>
          <p
            className="licood"
          >
            LONG:
             
            20
          </p>
        </div>
      </li>
    </a>
  </ul>
</div>
`;
